# A unit fraction contains 1 in the numerator. The decimal representation of the unit fractions with denominators 2 to 10 are given:

- 1/2	= 	0.5
- 1/3	= 	0.(3)
- 1/4	= 	0.25
- 1/5	= 	0.2
- 1/6	= 	0.1(6)
- 1/7	= 	0.(142857)
- 1/8	= 	0.125
- 1/9	= 	0.(1)
- 1/10	= 	0.1

# Where 0.1(6) means 0.166666..., and has a 1-digit recurring cycle. It can be seen that 1/7 has a 6-digit recurring cycle.

# Find the value of d < 1000 for which 1/d contains the longest recurring cycle in its decimal fraction part.

____

### Note: copied the idea from: https://www.mathblog.dk/project-euler-26-find-the-value-of-d-1000-for-which-1d-contains-the-longest-recurring-cycle/

# We remember the algorithm for division from elementary school as described below:

# Problem: calculate $\frac{1}{7}$

## $\frac{1}{7} = (0)(1) + (0.1)(\frac{10}{7})$

## $ = (0)(1) + (0.1)(1) + (0.01)(\frac{30}{7}) = (0)(1) + (0.1)(1) + (0.01)(4) + ... = 1.4286$

## More generally, the steps are:

### 1. Divide the numberator by the denominator, and keep the remainder

### 2. Multiply the remainder by 10 and divide by the denominator, keeping the new remainder

### 3. Multiply the quotient from part 2 by 0.1 and add to the total

### 4. Multiply the remainder from part 2 by 10 and divide ...


# From this, we can see that if we get a remainder twice it, there is a repeated decimal in the fraction and all decimals will be repeating in the same order thereafter

### This means we'll write our function to loop through the remainders and once one repeats we'll stop the counter

In [24]:
def count_cycle(d):
    list_remainders = []
    count = 0
    
    remainder = 1 % d
    
    while (remainder not in list_remainders)&(remainder!=0):
        count += 1
        list_remainders.append(remainder)
        remainder = remainder*10 % d
    return count

In [26]:
max_val = 0
max_count = 0

for d in range(1,1000):
    count = count_cycle(d)
    if count > max_count:
        max_count = count
        max_val = d

In [34]:
max_count, max_val

(982, 983)

# Answer: when $d=983$, the length of the recurring cycle is 982